ledger-service/http-json-perf/src/main/scala/com/daml/http/perf/Main.scala

```diff
@@ -116,51 +120,81 @@ object Main extends StrictLogging {
-      withJsonApiJdbcConfig(config.queryStoreIndex) { jsonApiJdbcConfig =>
-        withHttpService(
-          ledgerId.unwrap,
-          ledgerPort,
-          jsonApiJdbcConfig,
-          None,
-        ) { (uri, _, _, _) =>
-          runGatlingScenario(config, uri.authority.host.address, uri.authority.port)
-            .flatMap { case (exitCode, dir) =>
-              toFuture(generateReport(dir))
-                .map { _ =>
-                  logger.info(s"Report directory: ${dir.toAbsolutePath}")
-                  exitCode
-                }
-            }: Future[ExitCode]
+    for {
+      // For custom tokens we need to extract the expected ledger id, however for
+      // user tokens we can just choose any ledger id, because it doesn't contain any.
+      // If we have a user token we also extract it here so we can allocate the user later
+      // after we started the ledger.
+      userAndLedgerId <- HttpService
+        .decodeJwt(config.jwt)
+        .leftMap(_.message)
+        .flatMap { decodedJwt =>
+          CreateFromUserToken
+            .parseAndDecodeUserToken(decodedJwt)
+            .bimap(
+              error => s"This is neither a custom token nor a user token: ${error.message}",
+              {
+                case StandardJWTPayload(userId, _, _) =>
+                  (Some(userId), LedgerId("perf-runner"))
+                case CustomDamlJWTPayload(ledgerIdPayload, _, _, _, _, _, _) =>
+                  (None, LedgerId(ledgerIdPayload.getOrElse("perf-runner")))
+              },
+            )
         }
+        .fold((error: String) => Future.failed(new Exception(error)), Future.successful)
+      (userIdOpt, ledgerId) = userAndLedgerId
+      exitCode <- withLedger(config.dars, ledgerId.unwrap) { (ledgerPort, ledgerClient, _) =>
+        // For a user token to work the user has to be created beforehand.
+        for {
+          _ <-
+            userIdOpt.traverse { userId =>
+              ledgerClient.userManagementClient
+                .createUser(
+                  User(UserId.assertFromString(userId), None),
+                  List(CanActAs(Party.assertFromString("Alice"))),
+                )
+            }
+          res <- withJsonApiJdbcConfig(config.queryStoreIndex) { jsonApiJdbcConfig =>
+            withHttpService(
+              ledgerId.unwrap,
+              ledgerPort,
+              jsonApiJdbcConfig,
+              None,
+            ) { (uri, _, _, _) =>
+              runGatlingScenario(config, uri.authority.host.address, uri.authority.port)
+                .flatMap { case (exitCode, dir) =>
+                  toFuture(generateReport(dir))
+                    .map { _ =>
+                      logger.info(s"Report directory: ${dir.toAbsolutePath}")
+                      exitCode
+                    }
+                }: Future[ExitCode]
+            }
+          }
+        } yield res
       }
-    }
+    } yield exitCode
```

Comment:
There's a nested `for`-comprehension and a lot of indentation here. Breaking the steps into methods could improve the overall readability.